backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/pporder/api/impl/hu_pporder_issue_producer/CreatePickedIssueCommand.java

```diff
@@ -0,0 +1,220 @@
+		}
+
+		return candidate;
+	}
+
+	@Nullable
+	private I_PP_Order_Qty createIssueCandidateOrNull(
+			@NonNull final IHUContext huContext,
+			@NonNull final I_M_HU hu)
+	{
+		// Stop if we had to issue a fixed quantity and we already issued it
+		if (remainingQtyToIssue != null && remainingQtyToIssue.signum() <= 0)
+		{
+			return null;
+		}
+
+		final IHUProductStorage productStorage = retrieveProductStorage(huContext, hu);
+		if (productStorage == null)
+		{
+			return null;
+		}
+
+		// Actually create and save the candidate
+		final I_PP_Order_Qty candidate = createIssueCandidateOrNull(hu, productStorage);
+		if (candidate == null)
+		{
+			return null;
+		}
+		return candidate;
+	}
+
+	@Nullable
+	private IHUProductStorage retrieveProductStorage(
+			@NonNull final IHUContext huContext,
+			@NonNull final I_M_HU hu)
+	{
+		final List<IHUProductStorage> productStorages = huContext.getHUStorageFactory()
+				.getStorage(hu)
+				.getProductStorages();
+
+		// Empty HU
+		if (productStorages.isEmpty())
+		{
+			logger.warn("{}: Skip {} from issuing because its storage is empty", this, hu);
+			return null; // no candidate
+		}
+
+		if (productStorages.size() != 1)
+		{
+			throw HUException.ofAD_Message(MSG_IssuingHUWithMultipleProductsNotAllowed)
+					.setParameter("HU", hu)
+					.setParameter("ProductStorages", productStorages);
+		}
+
+		return productStorages.get(0);
+	}
+
+	@Nullable
+	private I_PP_Order_Qty createIssueCandidateOrNull(
+			@NonNull final I_M_HU hu,
+			@NonNull final IHUProductStorage productStorage)
+	{
+		try
+		{
+			final ProductId productId = productStorage.getProductId();
+
+			final Quantity qtyToIssue = calculateQtyToIssue(productStorage)
+					.switchToSourceIfMorePrecise();
+			if (qtyToIssue.isZero())
+			{
+				return null;
+			}
+
+			final I_PP_Order_Qty candidate = huPPOrderQtyDAO.save(CreateIssueCandidateRequest.builder()
+																		  .orderId(orderId)
+																		  //
+																		  .date(movementDate)
+																		  //
+																		  .locatorId(warehousesRepo.getLocatorIdByRepoIdOrNull(hu.getM_Locator_ID()))
+																		  .issueFromHUId(HuId.ofRepoId(hu.getM_HU_ID()))
+																		  .productId(productId)
+																		  //
+																		  .qtyToIssue(qtyToIssue)
+																		  //
+																		  .build());
+
+			return candidate;
+		}
+		catch (final RuntimeException rte)
+		{
+			throw AdempiereException.wrapIfNeeded(rte)
+					.appendParametersToMessage()
+					.setParameter("M_HU", hu);
+		}
+	}
+
+	/**
+	 * @return how much quantity to take "from" and issue it to given BOM line
+	 */
+	private Quantity calculateQtyToIssue(@NonNull final IHUProductStorage from)
+	{
+		if (remainingQtyToIssue != null)
+		{
+			final Quantity huStorageQty = from.getQty(remainingQtyToIssue.getUOM());
+			final Quantity qtyToIssue = huStorageQty.min(remainingQtyToIssue);
+
+			remainingQtyToIssue = remainingQtyToIssue.subtract(qtyToIssue);
```

Comment:
pls avoid side effects.
a method called `calculateQtyToIssue` it's not supposed to update the internal state